packages/regenesis-surgery/scripts/surgery.ts

```diff
@@ -0,0 +1,346 @@
+      OLD_UNISWAP_POOL_INITCODE_HASH
+    )
+    const newPoolAddress = getUniPoolAddress(
+      event.args.token0,
+      event.args.token1,
+      event.args.fee,
+      UNISWAP_V3_FACTORY_ADDRESS,
+      NEW_UNISWAP_POOL_INITCODE_HASH
+    )
+    // TODO: Throw an error if any of the pool addresses don't exist in the old state.
+    if (oldPoolAddress in dump.accounts) {
+      pools[oldPoolAddress] = {
+        newAddress: newPoolAddress,
+        token0: event.args.token0,
+        token1: event.args.token1,
+        fee: event.args.fee,
+      }
+    }
+  }
+
+  // Step 6. (UNISWAP) Fix the UniswapV3Factory `getPool` mapping.
+  for (const newPoolData of Object.values(pools)) {
+    // Fix the token0 => token1 => fee mapping
+    const oldSlotKey1 = getMappingKey(
+      [newPoolData.token0, newPoolData.token1, newPoolData.fee],
+      2
+    )
+    const newSlotKey1 = getMappingKey(
+      [newPoolData.token0, newPoolData.token1, newPoolData.fee],
+      5
+    )
+    dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[newSlotKey1] =
+      dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldSlotKey1]
+    delete dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldSlotKey1]
+
+    // Fix the token1 => token0 => fee mapping
+    const oldSlotKey2 = getMappingKey(
+      [newPoolData.token1, newPoolData.token0, newPoolData.fee],
+      2
+    )
+    const newSlotKey2 = getMappingKey(
+      [newPoolData.token1, newPoolData.token0, newPoolData.fee],
+      5
+    )
+    dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[newSlotKey2] =
+      dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldSlotKey2]
+    delete dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldSlotKey2]
+  }
+
+  // Step 7. (UNISWAP) Fix the NonfungiblePositionManager `poolId` mapping.
+  for (const [oldPoolAddress, newPoolData] of Object.entries(pools)) {
+    const oldSlotKey = getMappingKey([oldPoolAddress], 10)
+    const newSlotKey = getMappingKey([newPoolData.newAddress], 10)
+    dump.accounts[UNISWAP_NFPM_ADDRESS].storage[newSlotKey] =
+      dump.accounts[UNISWAP_NFPM_ADDRESS].storage[oldSlotKey]
+    delete dump.accounts[UNISWAP_NFPM_ADDRESS].storage[oldSlotKey]
+  }
+
+  // Step 8. (UNISWAP) Perform a final bruteforce step to find any remaining references to old addresses.
+  for (const [oldPoolAddress, newPoolData] of Object.entries(pools)) {
+    for (const [address, account] of Object.entries(dump.accounts)) {
+      if (account.storage === undefined) {
+        continue
+      }
+
+      // Check for any references to the pool address in storage values.
+      for (const [slotKey, slotValue] of Object.entries(account.storage)) {
+        if (slotValue.includes(oldPoolAddress)) {
+          // TODO: Figure out what to do here.
+          throw new Error(`found unexpected reference to pool address`)
+        }
+      }
+
+      // Check for single-level nested keys (i.e., address => xxxx).
+      for (let i = 0; i < 15; i++) {
```

Comment:
kelvin: this will have a very high ceiling

Reply:
"very high ceiling" = search more than the first 15 storage slots? If so makes sense—was going to say you should go way higher, especially if any mainnet contracts uses OZ upgradeable contracts since they reserve 50 empty slots at the end of each upgradeable contract

Reply:
Going to do 1000